src/foam/nanos/auth/ruler/EnsurePropertyOnCreateRule.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth.ruler',
+  name: 'EnsurePropertyOnCreateRule',
+  extends: 'foam.nanos.ruler.Rule',
+
+  documentation: `Rule to set the value of a specific property on the object if
+    it had not yet been set when created.
+
+    The rule is run after the object is put/created.
+
+    The predicate of EnsurePropertyOnCreateRule will check for
+    1. NEW_OBJ.spid == rule.spid
+    2. NEW_OBJ is an instance of the "targetClass"
+    3. The "propName" property on NEW_OBJ is not set
+
+    Then EnsurePropertyOnCreateRuleAction will set the object "propName"
+    property the value of the "propValue" and re-put the object back to the DAO.
+  `,
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.X',
+    'foam.nanos.ruler.predicate.IsInstancePredicate',
+    'foam.nanos.ruler.predicate.PropertyEQValue',
+    'foam.nanos.ruler.predicate.PropertyIsSetPredicate',
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    'id',
+    'name',
+    'documentation',
+    'enabled',
+    'priority',
+    'daoKey',
+    'ruleGroup',
+    {
+      class: 'Class',
+      name: 'targetClass',
+      section: 'basicInfo',
+      required: true
+    },
+    {
+      class: 'String',
+      name: 'propName',
+      section: 'basicInfo',
+      required: true
+    },
+    {
+      class: 'Object',
+      name: 'propValue',
+      section: 'basicInfo',
+      required: true
+    },
+    {
+      name: 'predicate',
+      transient: true,
+      visibility: 'HIDDEN',
+      javaGetter: `
+        var spidPredicate = new PropertyEQValue();
+        spidPredicate.setPropName("spid");
+        spidPredicate.setPropValue(getSpid());
+
+        var isInstanceOfPredicate = new IsInstancePredicate();
+        isInstanceOfPredicate.setOf(getTargetClass());
+
+        var propertyIsSetPredicate = new PropertyIsSetPredicate();
+        propertyIsSetPredicate.setPropName(getPropName());
+
+        return AND(
+          spidPredicate,
```

Comment:
What happens when the object is not ServiceProviderAware? 
You could use the ServiceProvderAwarePredicate.

Reply:
When object is not ServiceProviderAware the arg1 will be null since it doesn't have spid property and the arg2 is the spid of the current rule and is non-null therefore the rule predicate return false.

Reply:
ServiceProviderAwarePredicate will also check if the current user has permission to access the object which is what we don't want here. 

For example, there is rule1 which updates feeAccount of TransactionFeeRule if not set. We want rule1 to only apply to TransactionFeeRule with the same spid and not apply to TransactionFeeRule in other spids that user has access to.

Reply:
I don't understand.

Reply:
Well, ServiceProviderAwarePredicate will not check the current user permission if we explicitly set spid on the predicate. However, it still needs propertyInfos map which we don't have when constructing the predicate here.

Reply:
Is this used to support spid specific rules?

Reply:
It is to support spid specific rules.
```
// Auto-fill feeAccount for TransactionFeeRule on create
p({
  class: "foam.nanos.auth.ruler.EnsurePropertyOnCreateRule",
  id: "2e92ac64-e331-4152-a076-682b0c981c70",
  name: "Auto-fill TransactionFeeRule feeAccount",
  daoKey: "localRuleDAO",
  ruleGroup: "TransactionFeeRule",
  targetClass: "net.nanopay.tx.fee.TransactionFeeRule",
  propName: "feeAccount",
  propValue: 2,
  spid: "treviso",
  lifecycleState: 1
})
```
So this rule will update TransactionFeeRule.feeAccount to 2 if feeAccount is not set when the TransactionFeeRule is created in treviso spid.

Reply:
Fixed in https://github.com/foam-framework/foam2/pull/3956/commits/b893a287973bf560570cf2b0a5ae545ff7fe450f.